src/transformers/models/afmoe/modular_afmoe.py

```diff
@@ -0,0 +1,640 @@
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+
+def eager_attention_forward(
+    module: nn.Module,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attention_mask: Optional[torch.Tensor],
+    scaling: float,
+    dropout: float = 0.0,
+    **kwargs,
+):
+    key_states = repeat_kv(key, module.num_key_value_groups)
+    value_states = repeat_kv(value, module.num_key_value_groups)
+
+    attn_weights = torch.matmul(query, key_states.transpose(2, 3)) * scaling
+    if attention_mask is not None:
+        causal_mask = attention_mask[:, :, :, : key_states.shape[-2]]
+        attn_weights = attn_weights + causal_mask
+
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+    attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)
+    attn_output = torch.matmul(attn_weights, value_states)
+    attn_output = attn_output.transpose(1, 2).contiguous()
+
+    return attn_output, attn_weights
+
+
+class AfmoeRotaryEmbedding(LlamaRotaryEmbedding):
+    """
+    AFMoE reuses the standard Llama rotary embedding implementation.
+    """
+
+    pass
+
+
+class AfmoeRMSNorm(LlamaRMSNorm):
+    """
+    AFMoE shares the same RMSNorm definition as Llama.
+    """
+
+    pass
+
+
+class AfmoeMLP(LlamaMLP):
+    """
+    Wrapper around `LlamaMLP` that allows overriding the intermediate size per layer.
+    """
+
+    def __init__(self, config, intermediate_size: Optional[int] = None):
+        if intermediate_size is not None:
+            config = copy.copy(config)
+            config.intermediate_size = intermediate_size
+        if not hasattr(config, "mlp_bias"):
+            config.mlp_bias = False
+        super().__init__(config)
+
+
+class AfmoeTokenChoiceRouter(nn.Module):
+    """
+    Token-choice top-K router for MoE routing.
+
+    This router assigns each token to the top-K experts based on learned routing scores.
+    It supports both sigmoid and softmax scoring functions.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.top_k = config.num_experts_per_tok
+        self.num_experts = config.num_experts
+        self.score_func = config.score_func
+        self.route_norm = config.route_norm
+        self.route_scale = config.route_scale
+        self.gate = nn.Linear(config.hidden_size, config.num_experts, bias=False)
+
+    def forward(self, hidden_states, expert_bias: torch.Tensor | None):
+        _, _, hidden_dim = hidden_states.shape
+        hidden_states = hidden_states.view(-1, hidden_dim)
+
+        scores = self.gate(hidden_states)
+
+        # Apply scoring function in float32 for stability
+        if self.score_func == "sigmoid":
+            scores = torch.sigmoid(scores.to(torch.float32))
+        else:
+            scores = F.softmax(scores.to(torch.float32), dim=-1)
```

Comment:
in general we get rid of stuff that is not part of the released model, so if the released model uses only one of them, let's remove the other code pathes